_posts/2020-10-23-fine-grained-access-control-for-search.md

```diff
@@ -0,0 +1,99 @@
+
+Say your application uses standard HTTP endpoints to serve data to end users. So, a web user loads a webpage, the client-side Javascript fires off a HTTP request to the application endpoint for a specific type of data, the data is returned to the client-side Javascript and is rendered on the screen of the user. Your server code is doing some minor handling of the HTTP route and parameters, making a request to Elasticsearch based on this and manipulating the returned results to make it easily useable by the client-side Javascript. Pretty typical stuff.
+
+When your application becomes sufficiently complex, you’ll probably find the code is generally repetitious. You’ll be writing almost the same methods over and over - listen for a HTTP request, process it, ask Elasticsearch, return results, yet with small, subtle differences. Not only is this boring to write, it represents a larger, harder to maintain codebase. So, how does *access control* fit into all this?
+
+By varying the Open Distro for Elasticsearch users for types of different requests, you can change the data returned. Let’s take an e-commerce example. In this scenario you have documents that represent items - these documents can contain not only public facing fields - price, colour, material, description, but also private fields like cost, suppliers, and performance. Your employees may need to see all the information from behind a special login while customers should only see the public information. Of course, you could write very similar server code for each, vary the `_source` in the Query DSL and reimplement loads of logic. Alternatively, you can let Open Distro do this work for you and gain the ability to centrally control field availability.
+
+Let’s take this example - running this in query in Kibana Dev Tools (`<yourhost>:<yourport>/app/dev_tools#/console`) we will add a single item to the index `ecommerce-items`.
+
+
+```
+PUT ecommerce-items/_doc/brown-mug
+{
+    "title": "Brown Mug",
+    "description": "This mug features a large handle and a marbled, brown ceramic design.",
+    "price" : 5.99,
+    "private" : {
+    `    "cost" : 1,
+        "supplier" : "Mug World"
+    ` }
+}
+```
+
+
+No matter what happens, we don’t want our customers to be able see anything in the `private` object. We’ll use Open Distro’s built-in access control functionality to make this happen.
+
+## Creating a user & role
+
+First up in Kibana we need to create a user. This user will be used by the application for serving public information, so we’ll call it `public-item-user`. First, go to the “Security” item from the side menu then click “Internal users” in the secondary menu. Now, click on the “Create internal user” button. This will take you to `<yourhost>:<yourport>/app/opendistro_security#/users/create`.
```

Comment:
I generally recommend **Bold** for any GUI item that you have to click, but whatever you think is appropriate.

Reply:
I'll change that.

Reply:
Perfect. I'll use bold for clickables, double quotes for headings and labels, and monospaced for type-ins.